orangecontrib/text/widgets/owdocumentscorer.py

```diff
@@ -0,0 +1,389 @@
+    be applied to words.
+    """
+    # only transformers and normalizers preprocess on the word level
+    pps = [
+        pp
+        for pp in corpus.used_preprocessor.preprocessors
+        if isinstance(pp, (BaseTransformer, BaseNormalizer))
+    ]
+    for i, pp in enumerate(pps):
+        # TODO: _preprocess is protected make it public
+        words = [pp._preprocess(w) for w in words]
+        callback((i + 1) / len(pps))
+    return words
+
+
+def _run(
+    corpus: Corpus,
+    words: Table,
+    scoring_methods: List[str],
+    aggregation: str,
+    additional_params: dict,
+    state: TaskState,
+):
+    def callback(i: float):
+        state.set_progress_value(i * 100)
+        if state.is_interruption_requested():
+            raise Exception
+
+    cb_part = 1 / (len(scoring_methods) + 1)  # +1 for preprocessing
+
+    words = _preprocess_words(
+        corpus, words, wrap_callback(callback, end=cb_part)
+    )
+    for i, sm in enumerate(scoring_methods):
+        scoring_method = SCORING_METHODS[sm][1]
+        sig = signature(scoring_method)
+        add_params = {
+            k: v for k, v in additional_params.items() if k in sig.parameters
+        }
+        scs = scoring_method(
+            corpus,
+            words,
+            wrap_callback(
+                callback, start=(i + 1) * cb_part, end=(i + 2) * cb_part
+            ),
+            **add_params
+        )
+        scs = AGGREGATIONS[aggregation](scs, axis=1)
+        state.set_partial_result((sm, aggregation, scs))
+
+
+def set_cls_attributes(cls):
+    """
+    Class decorator that set widget settings dynamically for each scoring method
+    """
+    for i, name in enumerate(SCORING_METHODS):
+        setattr(cls, name, Setting(True if i == 0 else False))
+    for n, _ in ADDITIONAL_OPTIONS.values():
+        setattr(cls, n, Setting(0))
+    # ugly workaround - the problem is that metaclass new handles settings
+    # and this function is called afterwards - with this statement we handle
+    # settings again together with newly added settings
+    cls.settingsHandler = SettingsHandler.create(
+        cls, template=cls.settingsHandler
+    )
+    return cls
+
+
+@set_cls_attributes
+class OWDocumentScorer(OWWidget, ConcurrentWidgetMixin):
+    name = "Score documents"
+    description = ""
+    icon = "icons/CorpusViewer.svg"
+    priority = 500
+
+    auto_commit: bool = Setting(True)
+    aggregation: int = Setting(0)
+    # scoring method settings are generated by set_cls_attributes
+
+    class Inputs:
+        corpus = Input("Corpus", Corpus)
+        words = Input("Words", Table)
+
+    class Outputs:
+        corpus = Output("Corpus", Corpus)
+
+    class Warning(OWWidget.Warning):
+        missing_words = Msg("Provide words on the input")
+        missing_corpus = Msg("Provide corpus on the input")
+        corpus_not_normalized = Msg("Use preprocesses to normalize corpus.")
```

Comment:
Use *preprocessors* to normalize corpus. ?
How about Use Preprocess Text to normalize corpus. This would be in line with specific suggestions we have in Test and Score (for setting the target variable).